### PR TITLE
[metrics] Add quota_reserved_wait_time_seconds

### LINKED ISSUE #1961: [metrics] The admission_wait_time_seconds metric is misleading when AdmissionChecks are used
/kind bug
/kind feature

**What happened**:

The metric is incremented whenever a workload has QuotaReserved:  https://github.com/kubernetes-sigs/kueue/blob/b9c346a83d965d7849e4e27cf7e06e5ee1ad22df/pkg/scheduler/scheduler.go#L540

However, it may take arbitrarily long to admit the workload, but the metric says "The time between a Workload was created until it was admitted, per 'cluster_queue'".

**What you expected to happen**:

The metric should measure time to admission.

**How to reproduce it (as minimally and precisely as possible)**:

Create a cluster Queue with an admission check. 

**Proposed approach**:

The fix should contain two things:
- "fix" the metrics `admission_wait_time_seconds` so that it measures time to "Admitted" condition since creation time, or last eviction time
- "add" a new metric `quota_reserved_wait_time_seconds` which is measures what the current implementation is measuring

Regarding the  `admission_wait_time_seconds`  metric, the good starting point would be to move incrementing the metric under `IsAdmitted` [here](https://github.com/kubernetes-sigs/kueue/blob/b9c346a83d965d7849e4e27cf7e06e5ee1ad22df/pkg/scheduler/scheduler.go#L537-L539). However, there is another place which needs to be adjusted, around [here](https://github.com/kubernetes-sigs/kueue/blob/b9c346a83d965d7849e4e27cf7e06e5ee1ad22df/pkg/controller/core/workload_controller.go#L179-L182).

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind bug
/kind feature

#### What this PR does / why we need it:
Fixed metrics admitted_workloads_total, admission_wait_time_seconds.
Added metrics quota_reserved_workloads_total, quota_reserved_wait_time_seconds.

#### Which issue(s) this PR fixes:
Fixes #1961 

#### Does this PR introduce a user-facing change?
```release-note
Improve metrics relating to workload's quota reservation and admission
```

### CODE REVIEW

**Comment by trasc** [comment]
/release-note-edit ```Improve metrics relating to workload's quota reservation and admission```

**Comment by trasc** [comment]
/release-note-edit 
```
Improve metrics relating to workload's quota reservation and admission
```

**Comment by trasc** [comment]
/release-note-edit 

```release-note
Improve metrics relating to workload's quota reservation and admission
```

**Comment by trasc** [comment]
/release-note-edit

```
Improve metrics relating to workload's quota reservation and admission
```

File: test/integration/scheduler/podsready/scheduler_test.go [hunk 1]
```diff
@@ -254,6 +254,7 @@ var _ = ginkgo.Describe("SchedulerWithWaitForPodsReady", func() {
 			gomega.Expect(k8sClient.Create(ctx, prodWl)).Should(gomega.Succeed())
 			ginkgo.By("checking the 'prod' workload is admitted")
 			util.ExpectWorkloadsToHaveQuotaReservation(ctx, k8sClient, prodClusterQ.Name, prodWl)
+			util.ExpectQuotaReservedWorkloadsTotalMetric(prodClusterQ, 1)
```
Comment (mimowo) [inline test/integration/scheduler/podsready/scheduler_test.go:257]: It seems that in all cases the metric for reserved equals the metric for admitted. Please add an assert to a test where they differ, a good place might be `test/integration/controller/admissionchecks/provisioning/provisioning_test.go`.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -534,11 +534,15 @@ func (s *Scheduler) admit(ctx context.Context, e *entry, mustHaveChecks sets.Set
...
+			metrics.QuotaReservedWorkload(admission.ClusterQueue, waitTime)
+
 			if workload.IsAdmitted(newWorkload) {
 				s.recorder.Eventf(newWorkload, corev1.EventTypeNormal, "Admitted", "Admitted by ClusterQueue %v, wait time since reservation was 0s ", admission.ClusterQueue)
+				metrics.AdmittedWorkload(admission.ClusterQueue, 0)
 			}
-			metrics.AdmittedWorkload(admission.ClusterQueue, waitTime)
+
 			log.V(2).Info("Workload successfully admitted and assigned flavors", "assignments", admission.PodSetAssignments)
+
```
Comment (mimowo) [inline pkg/scheduler/scheduler.go:545]: nit: revert new empty line (here and above)

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -103,7 +119,7 @@ The label 'result' can have the following values:
 		prometheus.HistogramOpts{
 			Subsystem: constants.KueueName,
 			Name:      "admission_wait_time_seconds",
-			Help:      "The time between a Workload was created until it was admitted, per 'cluster_queue'",
+			Help:      "The time from when a workload got the quota reservation until admission, per 'cluster_queue'",
```
Comment (mimowo) [inline pkg/metrics/metrics.go:122]: This is a slightly different approach than in the Issue description, but I'm ok with either.

However, with the change of semantics I suggest a rename, maybe: `quota_reserved_to_admission_wait_time_seconds`

Another approach would be to call it `admission_checks_wait_time_seconds`, but then do not count 0 if there are no admission checks.

I'm slightly leaning towards `quota_reserved_to_admission_wait_time_seconds`. WDYT @alculquicondor @tenzen-y ?

File: site/content/en/docs/reference/metrics.md [hunk 4]
```diff
@@ -22,13 +22,15 @@ Use the following metrics to monitor the health of the kueue controllers:
 
 Use the following metrics to monitor the status of your ClusterQueues:
 
-| Metric name | Type | Description | Labels |
-| ----------- | ---- | ----------- | ------ |
-| `kueue_pending_workloads` | Gauge | The number of pending workloads. | `cluster_queue`: the name of the ClusterQueue<br> `status`: possible values are `active` or `inadmissible` |
-| `kueue_admitted_workloads_total` | Counter | The total number of admitted workloads. | `cluster_queue`: the name of the ClusterQueue |
-| `kueue_admission_wait_time_seconds` | Histogram | The time between a Workload was created until it was admitted. | `cluster_queue`: the name of the ClusterQueue |
+| Metric name | Type | Description                                                                    | Labels |
```
Comment (mimowo) [inline site/content/en/docs/reference/metrics.md:25]: Could you revert the formatting changes which don't result in the visual display of the table?
It is not immediate if these are just formatting or content changes.

**Review by mimowo — COMMENTED** [review]
Overall lgtm, couple of nits. I would also like to conclude on the new name for the new metric.

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 1 above)
Comment (mbobrovskyi) [reply test/integration/scheduler/podsready/scheduler_test.go:257]: Added assert on test/integration/scheduler/scheduler_test.go.

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (mbobrovskyi) [reply pkg/metrics/metrics.go:122]: Can I just rename metric to quota_reserved_to_admission_wait_time_seconds? Or should I need to change logic?

Comment (mimowo) [reply pkg/metrics/metrics.go:122]: `quota_reserved_to_admission_wait_time_seconds` sgtm. @alculquicondor any view?

File: test/util/util.go [hunk 5]
```diff
@@ -448,6 +448,15 @@ func ExpectAdmittedWorkloadsTotalMetric(cq *kueue.ClusterQueue, v int) {
 	}, Timeout, Interval).Should(gomega.Equal(v))
 }
 
+func ExpectQuotaReservedWorkloadsTotalMetric(cq *kueue.ClusterQueue, v int) {
+	metric := metrics.QuotaReservedWorkloadsTotal.WithLabelValues(cq.Name)
+	gomega.EventuallyWithOffset(1, func() int {
+		v, err := testutil.GetCounterMetricValue(metric)
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+		return int(v)
+	}, Timeout, Interval).Should(gomega.Equal(v))
```
Comment (tenzen-y) [inline test/util/util.go:457]:
```suggestion
	gomega.EventuallyWithOffset(1, func(g gomega.Gomega) {
		count, err := testutil.GetCounterMetricValue(metric)
		g.Expect(err).ToNot(gomega.HaveOccurred())
		g.Expect(int(count)).Should(gomega.Equal(v))
	}, Timeout, Interval).Should(gomega.Succeeded())
```

Improving debuggability would be better.

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/metrics/metrics.go:122]: sgtm. I would also not count a 0 if there are no admission checks

Comment (tenzen-y) [reply pkg/metrics/metrics.go:122]: SGTM

> I would also not count a 0 if there are no admission checks

This is a good call. I agree with this.

File: pkg/metrics/metrics.go [hunk 6]
```diff
@@ -103,7 +119,7 @@ The label 'result' can have the following values:
 		prometheus.HistogramOpts{
```
Comment (alculquicondor) [inline pkg/metrics/metrics.go:119]: Let's also add some buckets to both histograms.

The [DefBuckets](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#pkg-variables) only go up to 10s, which is definitely not enough in a queueing system.

Also, we don't really have resolution below 1s, so maybe we can use something like:

```
[]float64{1, 2.5, 5, 10, 20, 40, 80, 160, 320, 640, 1280, 2560, 5120, 10240}
```

Comment (mimowo) [reply pkg/metrics/metrics.go:119]: Great idea, I saw the cap by 10s already when looking into the metrics for the new scalability tests.

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (trasc) [reply pkg/metrics/metrics.go:122]: Then , maybe the name of the metric should be something like `admission_check_wait_time_seconds ` or keep two histograms. The transition from reserving to admitted happens with or without admissioncheckes .

In my opinion the way they are now, the counter and the number of recorded histogram records should be the same always.

Comment (mimowo) [reply pkg/metrics/metrics.go:122]: To follow up on the idea of `admission_check_wait_time_seconds` I opened: https://github.com/kubernetes-sigs/kueue/issues/1994.

Comment (alculquicondor) [reply pkg/metrics/metrics.go:123]: Can you summarize the full plan including all the involved metrics?

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 1 above)
Comment (mimowo) [reply test/integration/scheduler/podsready/scheduler_test.go:257]: thanks

File: site/content/en/docs/reference/metrics.md (hunk 4 above)
Comment (mimowo) [reply site/content/en/docs/reference/metrics.md]: this seems done, feel free to mark as resolved

File: pkg/scheduler/scheduler.go (hunk 2 above)
Comment (mimowo) [reply pkg/scheduler/scheduler.go]: feel free to mark as reasolved

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/metrics/metrics.go:123]: I agree with everything, except counting cases with no admission checks as zero in point 5.

Someone that wants to actually monitor `quota_reserved_to_admission_wait_time_seconds` doesn't care about queues that don't have admission checks.

Comment (mimowo) [reply pkg/metrics/metrics.go:123]: I see, but the name of the metric is sort of agnostic about the presence of admission checks. 

What would you say about calling it `admission_checks_wait_time_seconds`, and only counting when there are admission checks?

Comment (alculquicondor) [reply pkg/metrics/metrics.go:123]: SGTM

Comment (trasc) [reply pkg/metrics/metrics.go:123]: > 1. fix `admissionWaitTime` (`admission_wait_time_seconds`) - measures time to "Admitted" condition since creation time, or last eviction time

This is problematic the sense that when a quota reservation is done the eviction condition is altered and we are not able to recover that last evicted time, this is the reason why the event recording is done with the time from reservation to admission.

Comment (mimowo) [reply pkg/metrics/metrics.go:123]: Good point. So to implement this I see two ways:
1. have an extra condition like `Queued` with `LastTransitionTime` pointing to creation or last eviction, which is not updated on quota reservation, but on eviction
2. extra field in workload status to know how long quota reservation took, like `timeToReserveQuotaMs`

I would be leaning towards (1.) as more flexible.

Comment (mimowo) [reply pkg/metrics/metrics.go:123]: I like it

**Review by trasc — COMMENTED** [review]
/lgtm

File: test/integration/scheduler/scheduler_test.go [hunk 7]
```diff
@@ -378,6 +414,22 @@ var _ = ginkgo.Describe("Scheduler", func() {
 			})
 		})
 
+		ginkgo.It("Should admit workloads with admission checks", func() {
+			wl1 := testing.MakeWorkload("admission-check-wl1", ns.Name).
+				Queue(admissionCheckQueue.Name).
+				Request(corev1.ResourceCPU, "2").
+				Obj()
+
+			ginkgo.By("checking the first workload gets created and admitted", func() {
```
Comment (mimowo) [inline test/integration/scheduler/scheduler_test.go:423]:
```suggestion
			ginkgo.By("checking the first workload gets created and gets quota reserved", func() {
```

File: apis/kueue/v1beta1/workload_types.go [hunk 8]
```diff
@@ -302,6 +302,11 @@ const (
 	// more detailed information. The more detailed reasons should be prefixed
 	// by one of the "base" reasons.
 	WorkloadPreempted = "Preempted"
+
+	// WorkloadRequeued means that the Workload was requeued.
```
Comment (mimowo) [inline apis/kueue/v1beta1/workload_types.go:306]: Please make sure we have an integration test which checks that:
- `Requeued=True` when `QuotaReserved=True` and `Admitted=False`
- `Requeued=False` when the workload is evicted (IIUC this may still need implementation)

**Review by mimowo — COMMENTED** [review]
Almost lgtm

File: apis/kueue/v1beta1/workload_types.go (hunk 8 above)
Comment (mbobrovskyi) [reply apis/kueue/v1beta1/workload_types.go:306]: > Requeued=False when the workload is evicted (IIUC this may still need implementation)

Could you please explain this case? Is it possible to set Requeued to false?

Comment (mimowo) [reply apis/kueue/v1beta1/workload_types.go]: OTOH, setting it to `False` is not required to compute the metric, so maybe it does not need to be in this PR. WDYT @alculquicondor ?

Comment (mimowo) [reply apis/kueue/v1beta1/workload_types.go]: As discussed offline, we don't want to change the condition to False on admission.

File: pkg/controller/jobs/job/job_controller_test.go [hunk 9]
```diff
@@ -633,6 +633,12 @@ func TestReconciler(t *testing.T) {
 						Reason:  "Pending",
 						Message: "The workload is deactivated",
 					}).
+					Condition(metav1.Condition{
+						Type:    kueue.WorkloadRequeued,
+						Status:  metav1.ConditionTrue,
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller_test.go:638]: This is not correct. If the workload is deactivated, Requeued should be False.

File: apis/kueue/v1beta1/workload_types.go (hunk 8 above)
Comment (alculquicondor) [reply apis/kueue/v1beta1/workload_types.go]: The condition should transition to `False` in the following scenarios:
- When workload.spec.active=false. It should return to `True` when it's reactivated.
- When evicted due to WaitForPodsReady, the workload will temporarily be in backoff. During this time, `Requeued` should be false.

Although I think we can leave these for a follow up.

**Review by alculquicondor — COMMENTED** [review]
/approve
Leaving lgtm to @mimowo

**Comment by mimowo** [comment]
/lgtm 
Let's address the remaining comments ([test](https://github.com/kubernetes-sigs/kueue/pull/1977#discussion_r1574811084) and [transitions](https://github.com/kubernetes-sigs/kueue/pull/1977#discussion_r1574949456)) in follow up(s).

**Comment by mimowo** [comment]
> /lgtm Let's address the remaining comments ([test](https://github.com/kubernetes-sigs/kueue/pull/1977#discussion_r1574811084) and [transitions](https://github.com/kubernetes-sigs/kueue/pull/1977#discussion_r1574949456)) in follow up(s).

I opened the issue to track the remaining work for transitioning: https://github.com/kubernetes-sigs/kueue/issues/2038.

**Comment by mimowo** [comment]
/release-note-edit
```release-note
Improve metrics related to workload's quota reservation and admission:
- fix admission_wait_time_seconds - to measure the time to "Admitted" condition since creation time or last requeue (as opposed to the "QuotaReserved" condition as before)
- add quota_reserved_wait_time_seconds - measures time to "QuotaReserved" condition since creation time, or last eviction time
- add quota_reserved_workloads_total - counts the number of workloads that got admitted
- admission_checks_wait_time_seconds - measures the time to admit a workload with admission checks since quota reservation
- use longer buckets (up to 10240s) for histogram metrics: admission_wait_time_seconds, quota_reserved_wait_time_seconds, admission_checks_wait_time_seconds
```

**Comment by mimowo** [comment]
> @mimowo As I can see this PR, the semantics of the existing metric (`admitted_workloads_total ` and `admission_wait_time_seconds `) are reworked with breaking. So, shouldn't add the `ACTION REQUIRED` in the release note?

I don't think `admitted_workloads_total` changed, we just added `quota_reserved_workloads_total`.
Regarding `admission_wait_time_seconds` I would say it was fixed for workloads with admission checks. 
It does not seem to me user action is needed.

**Comment by tenzen-y** [comment]
> I don't think admitted_workloads_total changed, we just added quota_reserved_workloads_total.

Oh, I see. Thank you for the clarifications. I was thinking that the metric had been changed.

> Regarding admission_wait_time_seconds I would say it was fixed for workloads with admission checks.
It does not seem to me user action is needed.

That makes sense.
